### DISCUSSION TITLE
Map matching doesn't consider covariant generic type parameters

### DISCUSSION THREAD

**Discussion by lostllama** [LuckyPennySoftware/AutoMapper#3686] [Ideas]
### Source/destination types

```csharp
// Destination
public class Settings
{
	public string Test {get;set;}
}

// Source item
public interface ISetting
{
	string Name {get;set;}
	string Value {get;set;}
}

// Concrete source item
public class Setting : ISetting
{
	public string Name {get;set;}
	public string Value {get;set;}
}
```

### Mapping configuration

```csharp
var mc = new MapperConfiguration(cfg => {
    cfg.CreateMap<IEnumerable<ISetting>, Settings>()
        .ForMember(m => m.Test, o => o.MapFrom(c => c.FirstOrDefault(s => s.Name == "Test").Value)); 
});
		
var mapper = mc.CreateMapper();
```

### Version
AutoMapper 10.1.1
.NET 5

### Expected behavior

Mapping to succeed.

### Actual behavior

AutoMapper.AutoMapperMappingException: Missing type map configuration or unsupported mapping.
TypePair: Object -> Setting

### Steps to reproduce

```csharp
public static IEnumerable<ISetting> GetSettings()
{
	return new Setting[] { new Setting() { Name = "Test", Value = "Hello" } };
}

public static IEnumerable<ISetting> GetSettings2()
{
	return new ISetting[] { new Setting() { Name = "Test", Value = "Hello" } };
}

try
{
	Console.WriteLine();
	Console.WriteLine("As IEnumerable<ISetting>");
	Console.WriteLine();
	IEnumerable<ISetting> settings = GetSettings();
	Console.WriteLine($"Settings is {settings.GetType()}."); // prints Setting[]
	var result = mapper.Map<Settings>(settings);
	Console.WriteLine($"Success: Test is {result.Test}");
}
catch (AutoMapperMappingException e)
{
	var tp = (TypePair)e.Types;
	Console.WriteLine($"From {tp.SourceType} to {tp.DestinationType}"); // prints Object to Setting
	Console.WriteLine(e.ToString()); 
}

Console.WriteLine();
Console.WriteLine();

try
{
	Console.WriteLine();
	Console.WriteLine("As an array");
	Console.WriteLine();
	IEnumerable<ISetting> settings = GetSettings2();
	Console.WriteLine($"Settings is {settings.GetType()}."); // prints ISetting[]
	var result = mapper.Map<Settings>(settings);
	Console.WriteLine($"Success: Test is {result.Test}"); // prints Success: Test is Hello
}
catch (AutoMapperMappingException e)
{
	var tp = (TypePair)e.Types;
	Console.WriteLine($"From {tp.SourceType} to {tp.DestinationType}");
	Console.WriteLine(e.ToString());
}
```

This seems to be an issue related to `IEnumerable<out T>` being covariant.

[Fiddle](https://dotnetfiddle.net/mGbguS)

**Comment by lbargaoanu** [comment]
You're gonna have to create a map that actually matches. Or use `ISetting[]` instead of `Setting[]`.

**Reply by lostllama** [reply]
The issue here is that in the case where an `IEnumerable<ISetting>` holds a `Setting[]`, AutoMapper thinks that it's trying to map from `object` (or at least the error indicates it does) rather than, for example, `IEnumerable<ISetting>` or `Setting[]`. Why is that?

Curiously, if you're explicit about it, it works just fine:

`Settings s = mapper.Map<IEnumerable<ISetting>, Settings>(GetSettings());` works just fine.

It just seems a little inconsistent to me, and it seems strange that it can't recognise the source type and indicates that it's `object`.  Is this not a bug?

**Reply by lbargaoanu** [reply]
That's just a message. Try the [MyGet](https://docs.automapper.org/en/latest/The-MyGet-build.html) build. Specifying the source type works only because no map matches the runtime type.
